feat(productImport): Add filtering duplicate attributes

junajan opened the pull request:
#### Summary
Resolves #101 

#### Description
Functionality in this PR goes through all variants and filters out duplicate attributes (matched by name) and leaving only the first occurrence.

#### Todo

- Tests
    - [ ] Unit
    - [x] Integration
    - [ ] Acceptance
- [x] Documentation
<!-- Two persons should review a PR, don't forget to assign them. -->

butenkor commented on lib/product-import.coffee:
```diff
@@ -281,12 +282,46 @@ class ProductImport
+          msg = "Variant with SKU '#{variant.sku}' has duplicate attributes with name '#{attribute.name}'."
+          if @failOnDuplicateAttr
+            throw new Error(msg)
+          else
```
if @logOnDuplicateAttr
  logger...

junajan replied:
@butenkor updated here ebd0fa6

butenkor commented on lib/product-import.coffee:
```diff
@@ -281,12 +282,46 @@ class ProductImport
+            @logger.warn msg
+          return false
+        return true
+    variant
```
why do you return if it is not used?

junajan replied:
Updated here ebd0fa6

butenkor commented on lib/product-import.coffee:
```diff
@@ -281,12 +282,46 @@ class ProductImport
+      @_cleanVariantAttributes prodToProcess.masterVariant
+      prodToProcess.variants.forEach (variant) =>
+        @_cleanVariantAttributes variant
+    catch e
```
why do need try catch? You have all required NPE checks implemented already.

junajan replied:
Updated here ebd0fa6

butenkor commented on lib/product-import.coffee:
```diff
@@ -281,12 +282,46 @@ class ProductImport
     posts = _.map productsToProcess, (product) =>
       @_filterAttributes(product)
       .then (prodToProcess) =>
+        @_cleanDuplicateAttributes(prodToProcess)
```
this function does not do anything async so why it returns a promise?

junajan replied:
Chaged ebd0fa6

junajan commented:
Hello @butenkor - I updated PR according to your PR review here 405aa35.
Regarding the
> I seems that current error is not the same as res was before anymore -> has different structure. The idea of errorCallback was to pass response untouched back to caller so caller can decide how to handle/serialize/log it.

Originally the errorCallback was called not with an error object but with a Promise result object which has a `res.reason()` method for getting Promise result.

Also when I threw a `new error()` it was not logging an error correctly because of a [non iterable properties in error object](https://www.bennadel.com/blog/2874-error-object-properties-are-not-iterable-enumrable-in-node-js.htm) so the `fs.outputJsonSync` logged an empty object.

Therefore I added `serialize-error` and unified how the error is logged and what is passed to errorCallback function.

hisabimbola commented on test/integration.spec.coffee:
```diff
@@ -275,38 +276,39 @@ describe 'Product import integration tests', ->
         key: 'enum-new-key'
         label: 'enum-new-key'
 
-    newEnumKeyAttr2 =
```
I don't quite understand why we modified this test

junajan replied:
This test tried to save two attributes with the same name `sample_enum_attribute` and then read them both from the API. In this PR we filter out all duplicate attributes so only the first attribute got saved and thus this test was not working anymore.

hisabimbola commented on lib/product-import.coffee:
```diff
@@ -281,12 +286,38 @@ class ProductImport
+            throw new Error(msg)
+          else if @logOnDuplicateAttr
+            @logger.warn msg
+          return false
```
I don't think this is needed and the next line

junajan replied:
I refactored filter function here 293e85b so it is more readable.
